File: src/main/clients/openrgb/effects/DigitalRain.ts
import { Matrix } from '@main/clients/openrgb/client/classes/Matrix';
import { RGBColor } from '@main/clients/openrgb/client/classes/RGBColor';
import { Zone } from '@main/clients/openrgb/client/classes/Zone';
import { ZoneType } from '@main/clients/openrgb/client/interfaces/ZoneType';
import { AbstractEffect } from '@main/clients/openrgb/effects/base/AbstractEffect';

class DigitalRain extends AbstractEffect {
  public constructor() {
    super('Digital Rain', true);
  }

  private initializeMatrix(zoneStatus: Array<Array<number>>, zone: Zone): void {
    for (let i = 0; i < zone.matrix!.height; i++) {
      zoneStatus[i] = [];
      for (let j = 0; j < zone.matrix!.width; j++) {
        zoneStatus[i][j] = 0;
      }
    }
  }

  private initializeLinear(zoneStatus: Array<number>, zone: Zone): void {
    for (let i = 0; i < zone.ledsCount; i++) {
      zoneStatus[i] = 0;
    }
  }

  private decrementMatrix(zoneStatus: Array<Array<number>>, zone: Zone): void {
    for (let r = zone.matrix!.height - 1; r >= 0; r--) {
      for (let c = 0; c < zone.matrix!.width; c++) {
        if (r == 0) {
          if (zoneStatus[r][c] > 0) {
            zoneStatus[r][c] = Math.floor(zoneStatus[r][c] / 1.4);
          }
        } else {
          zoneStatus[r][c] = zoneStatus[r - 1][c];
        }
      }
    }
  }

  private decrementLinear(zoneStatus: Array<number>, zone: Zone): void {
    for (let r = zone.ledsCount - 1; r >= 0; r--) {
      if (r == 0) {
        if (zoneStatus[r] > 0) {
          zoneStatus[r] = Math.floor(zoneStatus[r] / 1.4);
        }
      } else {
        zoneStatus[r] = zoneStatus[r - 1];
      }
    }
  }

  private maxCount = Math.pow(2, 32);
  private toColorMatrix(
    zoneStatus: Array<Array<number>>,
    zone: Zone,
    matrix: Matrix
  ): Array<RGBColor> {
    const colors = Array(zone.ledsCount).fill(new RGBColor(0, 0, 0));
    for (let r = 0; r < zone.matrix!.height; r++) {
      for (let c = 0; c < zone.matrix!.width; c++) {
        if (matrix.keys[r][c] != undefined) {
          if (zoneStatus[r][c] == this.maxCount) {
            colors[matrix.keys[r][c]!] = new RGBColor(255, 255, 255);
          } else {
            colors[matrix.keys[r][c]!] = this.color.getDimmed(zoneStatus[r][c] / this.maxCount);
          }
        }
      }
    }
    return colors;
  }
  private toColorLinear(zoneStatus: Array<number>, zone: Zone): Array<RGBColor> {
    const colors = Array(zone.ledsCount).fill(new RGBColor(0, 0, 0));
    for (let r = 0; r < zone.ledsCount; r++) {
      colors[r] = this.color.getDimmed(zoneStatus[r] / this.maxCount);
    }
    return colors;
  }

  private matrixOnRate = 0;
  private getNextMatrix(status: Array<Array<number>>, matrix: Matrix): void {
    let countOff = 0;
    for (let c = 0; c < matrix.width; c++) {
      for (let r = 0; r < matrix.height; r++) {
        if (status[r][c] != 0) {
          break;
        } else if (r == matrix.height - 1) {
          countOff++;
        }
      }
    }
    if (countOff / matrix.width > this.matrixOnRate) {
      let next = -1;
      do {
        next = Math.floor(Math.random() * matrix.width);

        for (let r = 0; next >= 0 && r < matrix.height; r++) {
          if (status[r][next] != 0) {
            next = -1;
          } else {
            status[0][next] = this.maxCount;
          }
        }
      } while (next < 0);
    }
  }
  private getNextLinear(status: Array<number>, zone: Zone): void {
    let countOff = 0;
    for (let c = 0; c < zone.ledsCount; c++) {
      if (status[c] == 0) {
        countOff++;
      }
    }
    if (countOff / zone.ledsCount > this.matrixOnRate) {
      let next = -1;
      do {
        next = Math.floor(Math.random() * zone.ledsCount);

        if (status[next] != 0) {
          next = -1;
        } else {
          status[next] = this.maxCount;
        }
      } while (next < 0);
    }
  }

  protected async applyEffect(): Promise<void> {
    const zoneStatus: Array<Array<Array<number | Array<number>>>> = [];
    const colors: Array<Array<RGBColor>> = [];
    const promises: Array<Promise<void>> = [];
    this.devices.forEach((dev, id) => {
      promises.push(
        new Promise<void>((resolve) => {
          (async (): Promise<void> => {
            colors[id] = Array(dev.colors.length).fill(new RGBColor(0, 0, 0));
            this.setColors(dev, colors[id]);
            zoneStatus[id] = [];
            dev.zones.forEach((zone, iz) => {
              zoneStatus[id][iz] = [];

              switch (zone.type) {
                case ZoneType.MATRIX:
                  this.initializeMatrix(zoneStatus[id][iz] as Array<Array<number>>, zone);
                  break;
                case ZoneType.LINEAR:
                case ZoneType.SINGLE:
                  this.initializeLinear(zoneStatus[id][iz] as Array<number>, zone);
                  break;
              }
            });

            await this.sleep(Math.floor(Math.random() * 100));

            for (let iter = 0; this.isRunning; iter = (iter + 1) % 50) {
              let offset = 0;
              const finalColors: Array<RGBColor> = Array(dev.leds.length).fill(
                new RGBColor(0, 0, 0)
              );
              dev.zones.forEach((zone, iz) => {
                let colors: Array<RGBColor> | undefined = undefined;
                switch (zone.type) {
                  case ZoneType.MATRIX:
                    this.decrementMatrix(zoneStatus[id][iz] as Array<Array<number>>, zone);
                    if (iter % 7 == 0)
                      this.getNextMatrix(zoneStatus[id][iz] as Array<Array<number>>, zone.matrix!);
                    colors = this.toColorMatrix(
                      zoneStatus[id][iz] as Array<Array<number>>,
                      zone,
                      zone.matrix!
                    );
                    break;
                  case ZoneType.LINEAR:
                  case ZoneType.SINGLE:
                    this.decrementLinear(zoneStatus[id][iz] as Array<number>, zone);
                    if (iter % 7 == 0)
                      this.getNextLinear(zoneStatus[id][iz] as Array<number>, zone);
                    colors = this.toColorLinear(zoneStatus[id][iz] as Array<number>, zone);
                }
                for (let i = 0; i < colors!.length; i++) {
                  finalColors[offset + i] = colors![i];
                }
                offset += zone.ledsCount;
              });
              this.setColors(dev, finalColors);
              await this.sleep(75);
            }
            resolve();
          })();
        })
      );
    });

    await Promise.all(promises);
    this.hasFinished = true;
  }
}

export const digitalRain = new DigitalRain();

File: src/clients/openrgb/effects/digital_rain.py
from .base.abstract_effect import AbstractEffect
from ..sdk.orgb import Device
from ..sdk.utils import RGBColor, ZoneType
from ....utils.singleton import singleton

import math
import random
import threading


@singleton
class DigitalRain(AbstractEffect):
    def __init__(self):
        super().__init__("Digital rain", "#00FF00")
        self.decrement = 1.4
        self.max_count = int(pow(self.decrement, 32))
        self.matrix_on_rate = 0

    def initialize_matrix(self, zone_status, zone):
        for i in range(zone.mat_height):
            zone_status.append([])
            for j in range(zone.mat_width):
                zone_status[i].append(0)

    def initialize_linear(self, zone_status, zone):
        for i in range(len(zone.leds)):
            zone_status.append(0)

    def decrement_matrix(self, zone_status, zone):
        for r in range(zone.mat_height - 1, -1, -1):
            for c in range(zone.mat_width):
                if r == 0:
                    if zone_status[r][c] > 0:
                        zone_status[r][c] = math.floor(
                            zone_status[r][c] / self.decrement
                        )
                else:
                    zone_status[r][c] = zone_status[r - 1][c]

    def decrement_linear(self, zone_status, zone):
        for r in range(len(zone.leds) - 1, -1, -1):
            if r == 0:
                if zone_status[r] > 0:
                    zone_status[r] = math.floor(zone_status[r] / self.decrement)
            else:
                zone_status[r] = zone_status[r - 1]

    def to_color_matrix(self, zone_status, zone):
        colors = [RGBColor(0, 0, 0) for _ in range(len(zone.leds))]
        for r in range(zone.mat_height):
            for c in range(zone.mat_width):
                if zone.matrix_map[r][c] is not None:
                    if zone_status[r][c] == self.max_count:
                        colors[zone.matrix_map[r][c]] = RGBColor(255, 255, 255)
                    else:
                        colors[zone.matrix_map[r][c]] = self._color.dim(
                            zone_status[r][c] / self.max_count
                        )
        return colors

    def to_color_linear(self, zone_status, zone):
        colors = [RGBColor(0, 0, 0) for _ in range(len(zone.leds))]
        for r in range(len(zone.leds)):
            if zone_status[r] == self.max_count:
                colors[r] = RGBColor(255, 255, 255)
            else:
                colors[r] = self._color.dim(zone_status[r] / self.max_count)
        return colors

    def get_next_matrix(self, status, zone):
        count_off = 0
        for c in range(zone.mat_width):
            for r in range(zone.mat_height):
                if status[r][c] != 0:
                    break
                elif r == zone.mat_height - 1:
                    count_off += 1
        if count_off / zone.mat_width > self.matrix_on_rate:
            next_col = -1
            iters = 0
            while next_col < 0:
                iters += 1
                next_col = math.floor(random.random() * zone.mat_width)
                for r in range(zone.mat_height):
                    if status[r][next_col] != 0:
                        next_col = -1
                    else:
                        status[0][next_col] = self.max_count
                if iters > zone.mat_width * 5:
                    self.logger.error("Detected infinite loop")
                    self.logger.error(str(zone.device_id))
                    self.logger.error(str(zone.mat_width))
                    self.logger.error(str(zone.mat_height))
                    self.logger.error(str(zone.matrix_map))
                    self.logger.error(str(status))
                    break

    def get_next_linear(self, status, zone):
        count_off = sum(1 for c in status if c == 0)
        if count_off / len(zone.leds) > self.matrix_on_rate:
            next_idx = -1
            while next_idx < 0:
                next_idx = math.floor(random.random() * len(zone.leds))
                if status[next_idx] != 0:
                    next_idx = -1
                else:
                    status[next_idx] = self.max_count

    def apply_effect(self):
        zone_status = []
        colors = []
        threads = []

        def device_thread(dev: Device, dev_id):
            colors.append([RGBColor(0, 0, 0) for _ in dev.colors])
            self.set_colors(dev, colors[dev_id])
            zone_status.append([])

            for iz, zone in enumerate(dev.zones):
                zone_status[dev_id].append([])

                if zone.type == ZoneType.MATRIX:
                    self.initialize_matrix(zone_status[dev_id][iz], zone)
                elif zone.type in (ZoneType.LINEAR, ZoneType.SINGLE):
                    self.initialize_linear(zone_status[dev_id][iz], zone)

            self.sleep(random.randint(0, 100) / 1000)

            iter_count = 0
            while self.is_running:
                offset = 0
                final_colors = [RGBColor(0, 0, 0) for _ in dev.colors]

                for iz, zone in enumerate(dev.zones):
                    can_add = (iter_count % 7) == 0
                    zone_colors = None

                    if zone.type == ZoneType.MATRIX:
                        self.decrement_matrix(zone_status[dev_id][iz], zone)
                        if can_add:
                            self.get_next_matrix(zone_status[dev_id][iz], zone)
                        zone_colors = self.to_color_matrix(
                            zone_status[dev_id][iz], zone
                        )
                    elif zone.type in (ZoneType.LINEAR, ZoneType.SINGLE):
                        self.decrement_linear(zone_status[dev_id][iz], zone)
                        if can_add:
                            self.get_next_linear(zone_status[dev_id][iz], zone)
                        zone_colors = self.to_color_linear(
                            zone_status[dev_id][iz], zone
                        )

                    for i, color in enumerate(zone_colors):
                        final_colors[offset + i] = color
                    offset += len(zone.leds)

                self.set_colors(dev, final_colors)
                self.sleep(0.05 + random.randint(0, 50) / 1000)
                iter_count = (iter_count + 1) % 50

        for dev_id, dev in enumerate(self.devices):
            thread = threading.Thread(
                name=f"DigitalRain-dev-{dev_id}",
                target=device_thread,
                args=(dev, dev_id),
            )
            threads.append(thread)
            thread.start()

        for thread in threads:
            thread.join()


digital_rain = DigitalRain()
